[Android][libass] Enable fontconfig for libass to support all language scripts in subtitles

naseef opened the pull request:
## Description
This PR enables fontconfig support for libass on Android, providing automatic font fallback for all language scripts without requiring manual user configuration.

## Motivation and context

**Problem:** On Android, Kodi currently has **no font fallback mechanism** for subtitle rendering. When subtitles contain non-Latin scripts (Malayalam, Tamil, Arabic, Chinese, Japanese, Korean, Hebrew, etc.), characters are displayed as **empty boxes (□)** instead of proper glyphs.

**Current workaround:** 
Users must:
1. Download appropriate font files manually
2. Copy them to Kodi's userdata folder
3. Navigate through settings to select the font
4. Repeat for each language/script they need

This is **extremely difficult** for novice users, especially on:
- **Android TV boxes** (where Kodi is heavily used) - limited file management capabilities
- **Fire TV devices** - no easy file access

## How has this been tested?
- Devices:
  - Android 15 Phone
  - Amazon Fire TV Stick (Fire OS 7.7.0.2 - Android 9 Pie)
- Kodi version: Built from master branch with this PR
- Test subtitle files: Malayalam (.srt)
## What is the effect on users?
**Before this PR:**
- Non-Latin subtitle characters appear as boxes (□) on Android
- Users must manually download and install fonts
- Complex setup process unsuitable for TV box users
- Poor user experience for international users

**After this PR:**
- **Zero-configuration font fallback** - all languages work out of the box
- Subtitles in Malayalam, Tamil, Chinese, Japanese, Korean, Arabic, Hebrew, Thai, Hindi, and 100+ other scripts render correctly
- Uses Android's built-in font family (already on device)
- Seamless experience matching Windows/macOS/Linux platforms
- Especially beneficial for **Android TV box users** who have limited file management options

**User impact: HIGH** - Fixes a fundamental usability issue affecting all international Android users

## Screenshots (if appropriate):

**Before (without this PR and fresh install):**

![before](https://github.com/user-attachments/assets/b1dc09a9-a9d3-4376-989c-bc40fd3cf73f)

**After (with this PR and fresh install):**
![after](https://github.com/user-attachments/assets/af6a9f51-5c96-4d94-947d-f1fc8fb9dc2f)
## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [x] I have added tests to cover my change
- [x] All new and existing tests passed

CastagnaIT commented on xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:
```diff
@@ -92,11 +92,93 @@ void CDVDSubtitlesLibass::Configure()
+
+#if defined(TARGET_ANDROID)
+  // Use fontconfig with custom config pointing to Android system fonts.
+  // Fontconfig provides automatic font fallback for all scripts (Malayalam, Tamil, Chinese, etc.).
```
> // Use fontconfig with custom config pointing to Android system fonts.

All following code is clearly a _workaround / hack_ since Libass does not currently support native Android "fontconfig", this must be clearly specified, so instead to remove in full the original comment, it must be readapted

> // Fontconfig provides automatic font fallback for all scripts (Malayalam, Tamil, Chinese, etc.).

this comment is not required, is an intrinsic feature

CastagnaIT commented on xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:
```diff
@@ -92,11 +92,93 @@ void CDVDSubtitlesLibass::Configure()
+                              "  <dir>/system/fonts</dir>\n"
+                              "  <dir>/product/fonts</dir>\n"
+                              "  <dir>/system_ext/fonts</dir>\n"
+                              "  <dir>/vendor/fonts</dir>\n"
```
Some of these paths may not exist because brands do not always follow standards,
if there is any documentation regarding these paths, it would be best to mention it

naseef replied:
@CastagnaIT Thank you for the review! Here's the documentation for the font directories:

  **Confirmed in Android documentation and source code:**

  1. **`/system/fonts`** and **`/product/fonts`**
     - Documented in: [Android Custom Font Fallback](https://source.android.com/docs/core/fonts/custom-font-fallback)
     - Source code: [system_fonts.cpp](https://android.googlesource.com/platform/frameworks/base.git/+/master/native/android/system_fonts.cpp)

  2. **`/system_ext/fonts`**
     - OEM-specific extension to system partition
     - Documented in: [Shared System Image - system_ext partition](https://source.android.com/docs/core/architecture/partitions/shared-system-image#sys-ext-partition)
     - Used by manufacturers for OEM customizations without modifying `/system`

  3. **`/vendor/fonts`**
     - Vendor-specific fonts partition
     - Used by some vendors (example: [VLC source code](https://github.com/videolan/vlc/blob/05f437f98ec2b830315fe11e837049c227bf3662/modules/text_renderer/freetype/fonts/android.c#L49))

 Should I add these documentation references as code comments in the implementation?

CastagnaIT replied:
i made some searches about these paths:
`/system_ext/fonts`
`/vendor/fonts`

there is nothing explained on linked documentation and i found nothing else on web
linked VLC source only point the xml thats seem different i have not seen the "fonts" subfolder

after a detailed discussion with AI it was unable to provide me with any documentation or confirmation on other websites, the only thing that AI said is the “possibility” that these two paths are used by vendors but are undocumented

by checking on my two android devices (Motorola, Nvidia Shield) these two paths dont exists

so verify and prove the existence of these paths in the brands firmwares and mention what they are
or delete these two paths and stick to what the Android docs specifies

naseef replied:
You are correct that **`/vendor/fonts`** is not used. However, **`/system_ext/fonts`** is present on some Android devices. According to the [Android documentation](https://source.android.com/docs/core/architecture/partitions/shared-system-image#sys-ext-partition), the `system_ext` partition serves as an extension of the `system` partition, and you can see an example of this directory use [here in an android ROM dump](https://dumps.tadiphone.dev/dumps/oneplus/op555bl1/-/tree/b643a388157563387f75d004602cad364b185cc0/system_ext/fonts).

That said, neither path exists on my own device. Since `fontconfig` simply ignores non-existent directories, my suggestion is to **keep `/system_ext/fonts`** for compatibility with devices that do use it and **remove `/vendor/fonts`**.

What do you think?

CastagnaIT reviewed (commented):
I haven't tested anything, i have only take a look part of the code
this could be a temporary workaround until Libass implements something native

Also many “const” are missing in the code that should be added

naseef committed:
[Android] Enable fontconfig for libass to support all language scripts in subtitles

Currently, only user-specified fonts and Kodi's default skin fonts are available
for subtitle rendering on Android. This means subtitles with non-Latin scripts
(Malayalam, Tamil, Arabic, Chinese, Japanese, Korean, Hebrew, etc.) display as
empty boxes (□) unless users manually download and install appropriate fonts.

This commit enables fontconfig support for libass on Android by creating a custom
fontconfig configuration pointing to Android system fonts, allowing automatic
discovery and use of fonts which contain glyphs for virtually all languages.

This is a temporary workaround until libass implements native Android font provider support.

Changes:
- Use CXBMCTinyXML2 for safe XML generation of fontconfig configuration file
- Configure fontconfig to scan Android system font directories

This provides zero-configuration font fallback for all language scripts.

naseef commented on xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:
```diff
@@ -92,11 +92,93 @@ void CDVDSubtitlesLibass::Configure()
+                              "  <cachedir>" +
+                              escapedCacheDir +
+                              "</cachedir>\n"
+                              "</fontconfig>\n";
```
Code updated to use CXBMCTinyXML2

naseef commented:
> I haven't tested anything, i have only take a look part of the code this could be a temporary workaround until Libass implements something native
> 
> Also many “const” are missing in the code that should be added

Thanks for the review. Suggestion incorporated, Please check

CastagnaIT commented:
@ksooo about claude point : `1. Potential race condition`
we should really consider to protect all code in the `CDVDSubtitlesLibass::Configure` method with a mutex due to file/directory operations uses? fwik only one libass instance can be created when in playback

CastagnaIT commented on xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:
```diff
@@ -92,10 +93,81 @@ void CDVDSubtitlesLibass::Configure()
+  // Create fontconfig cache directory
+  const std::string fontconfigCacheDir =
+      CSpecialProtocol::TranslatePath("special://temp/fontconfig");
+  XFILE::CDirectory::Create(fontconfigCacheDir);
```
```
if (!XFILE::CDirectory::Exists(fontconfigCacheDir))
  XFILE::CDirectory::Create(fontconfigCacheDir);
```

CastagnaIT reviewed (approved):
i dont have other requests
LGTM
